### DISCUSSION TITLE
`compilation.getAssets()` results in `DeprecationWarning`

### DISCUSSION THREAD

**Discussion by ajmeese7** [webpack/webpack#16231] [Q&A]
I have the following code as part of a Webpack plugin:

```js
compiler.hooks.emit.tap(pluginName, (compilation) => {
	const icons = compilation.getAssets().map(
		(asset) => asset.name.split("/").pop()
	);

	...
});
```

It results in the following warning:

```
(node:505721) [DEP_WEBPACK_COMPILATION_ASSETS] DeprecationWarning: Compilation.assets will be frozen in future, all modifications are deprecated.
BREAKING CHANGE: No more changes should happen to Compilation.assets after sealing the Compilation.
        Do changes to assets earlier, e. g. in Compilation.hooks.processAssets.
        Make sure to select an appropriate stage from Compilation.PROCESS_ASSETS_STAGE_*.
(Use `node --trace-deprecation ...` to show where the warning was created)
```

Is this a mistake on my part or a Webpack bug? I am not modifying any of the assets, so I have a feeling that this may be a "false positive" warning. I wanted to ask here first before filing a bug report.

Thanks for your thoughts!

**Comment by vankop** [comment]
to avoid warning you can set `experiments.backCompat = false`

**Reply by ajmeese7** [reply]
Will give that a look when I get back from work, thanks for your answer!

**Reply by ajmeese7** [reply]
When I enabled that flag, I started getting the following error:

```text
[webpack-cli] TypeError: Cannot add property ../metadata.json, object is not extensible
```

The only place where "../metadata.json" is used is here, where it is the value of `relativePath`:

```js
const json = JSON.stringify(newMetadata, null, 2);
const relativePath = path.relative(compiler.outputPath, metadataFile);
compilation.emitAsset(relativePath, new RawSource(json));
```

Any ideas on how to fix that?

**Reply by alexander-akait** [reply]
> I am not modifying any of the assets, so I have a feeling that this may be a "false positive" warning.

But you write a new file, so you have this warhing, please migrate on new hooks, we have the `stage` option for these cases
